app/models/reducer.rb

```diff
@@ -43,67 +45,127 @@ def self.of_type(type)
+      reduced_extracts.select { |reduction| reduction&.data }
     end
 
     light.run
   end
 
-  def get_reduction(reduction_fetcher, group_key)
-    if running_reduction?
-      reduction_fetcher.retrieve_in_place(reducer_key: key, subgroup: group_key).tap do |r|
-        r.data ||= {}
-        r.store ||= {}
-      end
+  # for each extract, try to find a relevant reduction if relevant
+  # reductions are configured
+  #
+  # relevant_reductions are any previously reduced user or subject reductions
+  # that are required by this reducer to properly calculate
+  def augment_extracts(extracts)
+    relevant_reductions = get_relevant_reductions(extracts)
+    return extracts if relevant_reductions.empty?
+
+    extracts.each do |ex|
+      ex.relevant_reduction = if reduce_by_subject?
+          # find will only return the first match
+          relevant_reductions.find { |rr| rr.user_id == ex.user_id }
+        elsif reduce_by_user?
+          # find will only return the first match
+          relevant_reductions.find { |rr| rr.subject_id == ex.subject_id }
+        else
+          raise NotImplementedError.new 'This reduction topic is not supported'
+        end
+    end
+  end
+
+  # load all reductions that might be relevant to any of the extracts that are
+  # being reduced
+  def get_relevant_reductions(extracts)
+    return [] if user_reducer_keys.blank? && subject_reducer_keys.blank?
+
+    if reduce_by_subject?
+      UserReduction.where(user_id: extracts.map(&:user_id), reducible: reducible, reducer_key: user_reducer_keys)
+    elsif reduce_by_user?
+      SubjectReduction.where(subject_id: extracts.map(&:subject_id), reducible: reducible, reducer_key: subject_reducer_keys)
     else
-      reduction_fetcher.retrieve(reducer_key: key, subgroup: group_key).tap do |r|
-        # send empty data and store unless we're in running reduction mode
-        r.data = {}
-        r.store = {}
-      end
+      raise NotImplementedError.new 'This reduction mode is not supported'
     end
   end
 
+  # apply extract filters and deduplicate extracts (do not allow a running reduction
+  # to contain the same extract twice)
   def filter_extracts(extracts, reduction)
-    extracts = extracts.reject{ |extract| reduction.extract_ids.include? extract.id }
-    extract_filter.apply(extracts)
-  end
+    return extracts if extracts.blank?
 
-  def associate_extracts(reduction, extracts)
-    extracts.each do |extract|
-      reduction.extracts << extract
+    extracts = extract_filter.apply(extracts)
+    if running_reduction?
+      extracts = extracts.reject { |extract| reduction.extract_ids.include? extract.id }
     end
+
+    extracts
   end
 
-  def reduce_into(extracts, reduction)
-    raise NotImplementedError
+  # given an array of reductions to be updated, find or create one with the
+  # appropriate keys
+  def get_group_reduction(reductions, group_key)
+    requested_reduction = reductions.find{ |reduction| reduction.subgroup == group_key }
```

Comment:
This was the original bug, previously this behaviour would modify the existing reduction to wipe `data` &  `store` if in default mode. However the buggy PR returns the reduction as is no matter if running mode or not.